### PR TITLE
Fixed #28849 -- Fixed referenced table and column rename on SQLite.

### PR DESCRIPTION
**Pull request by charettes**
https://code.djangoproject.com/ticket/28849

### CODE REVIEW

**Comment by timgraham** [comment]
In [the SQLite docs](https://sqlite.org/lang_altertable.html), I found:
> If foreign key constraints are enabled when a table is renamed, then any REFERENCES clauses in any table (either the table being renamed or some other table) that refer to the table being renamed are modified to refer to the renamed table by its new name. 

The test passes if disabling of constraint checking in sqlite's `DatabaseSchemaEditor.__init__()` is removed, but this introduces other failures.

I tried this patch but it doesn't change behavior. I guess calls to enable/disable foreign keys can't be nested.
```diff
diff --git a/django/db/backends/base/base.py b/django/db/backends/base/base.py
index 468eb16..eed6564 100644
--- a/django/db/backends/base/base.py
+++ b/django/db/backends/base/base.py
@@ -451,6 +451,15 @@ class BaseDatabaseWrapper:
             if disabled:
                 self.enable_constraint_checking()
 
+    @contextmanager
+    def constraint_checks_enabled(self):
+        """Enable foreign key constraint checking."""
+        self.enable_constraint_checking()
+        try:
+            yield
+        finally:
+            self.disable_constraint_checking()
+
     def disable_constraint_checking(self):
         """
         Backends can implement as needed to temporarily disable foreign key
diff --git a/django/db/backends/sqlite3/schema.py b/django/db/backends/sqlite3/schema.py
index cbb018e..9130b6a 100644
--- a/django/db/backends/sqlite3/schema.py
+++ b/django/db/backends/sqlite3/schema.py
@@ -53,6 +53,10 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
         else:
             raise ValueError("Cannot quote parameter value %r of type %s" % (value, type(value)))
 
+    def alter_db_table(self, model, old_db_table, new_db_table):
+        with self.connection.constraint_checks_enabled():
+            super().alter_db_table(model, old_db_table, new_db_table)
+
     def _remake_table(self, model, create_field=None, delete_field=None, alter_field=None):
         """
         Shortcut to transform a model from old_model into new_model
```

**Comment by charettes** [comment]
> We'll still have to figure out how to deal with the to_field rename case thought as this requires a table rebuild.

It looks like the same documentation section (second procedure) documents a way to perform a column rename without a table rebuild so that might just work.

**Comment by charettes** [comment]
@timgraham alright so I did a bit of investigation here and it looks like this will be hard to solve.

If we remove the constraint disabling in [`DatabaseSchemaEditor.__enter__`](https://github.com/django/django/blob/4d3b8e199ec00a0aba111637a25f5abcaf3d6dfc/django/db/backends/sqlite3/schema.py#L20-L22) the tests fail for the same reason the new test pass; references are repointed appropriately when foreign key constraints are enabled. The failing tests are all exercising `_remake_table` which rename the existing table to `table_name_old` and the references are then updated to point to the `table_name_old` table which doesn't exist anymore on commit resulting in errors similar to

```python
django.db.utils.OperationalError: no such table: main.test_authors_author__old
```

SQLite documentation mentions that it should be possible to [repoint references by updating rows of `sqlite_master` table manually to alter the SQL used to create the table](https://sqlite.org/lang_altertable.html#otheralter). Unfortunately I couldn't get it to work. While `.schema` reports the table was updated and I followed the instruction it keeps complaining that the old table doesn't exist or that the foreign key still exist.

```sql
SQLite version 3.11.0 2016-02-15 17:29:24
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
sqlite> PRAGMA foreign_keys = on;
sqlite> CREATE TABLE foo ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT);
sqlite> CREATE TABLE bar ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "foo_id" integer NOT NULL REFERENCES "foo" ("id") DEFERRABLE INITIALLY DEFERRED);
sqlite> BEGIN;
sqlite> PRAGMA schema_version;
2
sqlite> PRAGMA writable_schema=1;
sqlite> UPDATE sqlite_master SET sql = replace(sql, 'NOT NULL REFERENCES "foo" ("id") DEFERRABLE INITIALLY DEFERRED', '');
sqlite> PRAGMA schema_version=3;
sqlite> PRAGMA writable_schema=0;
sqlite> PRAGMA integrity_check;
ok
sqlite> COMMIT;
sqlite> INSERT INTO bar(foo_id) VALUES(42);
Error: FOREIGN KEY constraint failed
sqlite> PRAGMA foreign_keys = off;
sqlite> PRAGMA foreign_keys = on;
sqlite> INSERT INTO bar(foo_id) VALUES(42);
Error: FOREIGN KEY constraint failed
```

Here's the three solutions I have to propose.

1. Set SQLite `can_rollback_ddl` to false which will prevent the migration from being wrapped in a transaction and allow us to freely enable and disable foreign keys.
2. Raise an exception in SQLite's `rename_db_table` and `rename_column` if the table/column is referenced by another table, foreign key constraints are disabled (default), and we're currently in a transaction (default). The exception message could mention that the migration's `atomic` attribute must be set to `False` to allow the rename operation to proceed appropriately.
3. Disable foreign key support for SQLite by default until it matures a bit more and mention it can be turned but this might cause issues with migration. The fact foreign key constraints can't be enabled or disabled in a transaction is really a bummer.

I'd favor 2. as it's the less invasive of the three options and allows shipping SQLite foreign support and avoids dropping support for atomic migrations in most cases.

**Comment by charettes** [comment]
After looking around a bit more it looks the [`UPDATE sqlite_master`  approach requires either a connection close/reconnect or a vacuum to reload the table definition](https://stackoverflow.com/a/4820137/70191) so it's a dead end in our case because this cannot be performed within a transaction.

**Comment by charettes** [comment]
@timgraham, finally managed to come up with a solution for both table and column rename. The only part missing is

1. Testing the error messages raised when a referenced table or column is attempted to be renamed in a transaction on SQLite.
2. Adjust the SQLite release notes related to foreign key support to mention that such operations are now required to be run in non-atomic migrations.

File: django/db/backends/sqlite3/schema.py [hunk 1]
```diff
@@ -53,6 +55,63 @@ def quote_value(self, value):
...
+                    new_column_name = new_field.get_attname_column()[1]
+                    search = references_template % old_column_name
+                    replacement = references_template % new_column_name
+                    cursor.execute('UPDATE sqlite_master SET sql = replace(sql, %s, %s)', (search, replacement))
+                    cursor.execute('PRAGMA schema_version = %d' % (schema_version + 1))
+                    cursor.execute('PRAGMA writable_schema = 0')
+                    # Raise an exception if the sqlite_master updates broke
+                    # anything to make sure the transaction is rolled back.
+                    if cursor.execute('PRAGMA integrity_check').fetchone()[0] != 'ok':
+                        raise Exception('Failed to repoint references from "%s"."%s" to "%s"."%s".' % (
```
Comment (charettes) [inline django/db/backends/sqlite3/schema.py]: I think we might want to raise an `IntegrityError` with the output of `PRAGMA integrity_check` here instead. Not sure if `PRAGMA integrity_check` will raise an error by itself, you can test that out by changing `replacement` to invalid SQL.

**Comment by ramiro** [comment]
> Testing the error messages raised when a referenced table or column is attempted to be renamed in a transaction on SQLite.

Did this here: https://github.com/django/django/compare/05bf9dbafabd17aa08da585d071c7f0771f0f2a6...ramiro:t28849-pr9389-details-and-tests

**Pull request merged**